### PR TITLE
3583 upgrade to ghc 9.2.7

### LINKED ISSUE #3583: Upgrade haskell-backend to ghc-9.2.7

### PR DESCRIPTION
**Pull request by jberthold**
Fixes #3583 

### Scope:

* Upgrades to lts-20.20 and ghc-9.2.7
* Removes all non-flake `nix` components
* Removes the "performance" measurement (which was not very useful any more)
* upgrades the Linux builds to ubuntu-22.04
* upgrades `hlint` to version `3.5` (minor adjustments required)
* integrates the formatting pass into the `Test` workflow 
   - for now it just prints a diff after formatting
   - later it should push a formatting commit and abort the workflow (exit with error) (next PR)

### CODE REVIEW

File: flake.lock
Comment (goodlyrottenapple) [inline flake.lock:1]: the lock file has to be modified automatically by nix, specifically via

```
nix flake lock --update-input haskell-nix
```

**Comment by jberthold** [comment]
The build seems to reliably fail with `No space left on device` now... :disappointed: 
https://github.com/runtimeverification/haskell-backend/actions/runs/4958152308/jobs/8903259950#step:5:10136
Maybe we are indeed going over the runner's resources? However, we seem to be building `kore` _twice_, and the build that fails seems to use `ghc-9.2.5` ???

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -90,11 +153,15 @@ jobs:
           skipPush: true
 
       - name: Run integration tests
-        run: nix-build test.nix
+        run: |
+          nix develop github:runtimeverification/k/sam/nix-kore-integration-shell#kore-integration-tests --override-input haskell-backend . --update-input haskell-backend
+          cd test
+          make -j8 -l8 --output-sync test
```
Comment (goodlyrottenapple) [inline .github/workflows/test.yml:159]: change to 

```bash
cd test
nix develop github:runtimeverification/k/sam/nix-kore-integration-shell#kore-integration-tests --override-input haskell-backend ./.. --update-input haskell-backend .#devops -c bash -c "make -j8 -l8 --output-sync test"
```

Comment (jberthold) [reply .github/workflows/test.yml:159]:
```suggestion
          nix develop github:runtimeverification/k/sam/nix-kore-integration-shell#kore-integration-tests \
                --override-input haskell-backend . --update-input haskell-backend \
                --command bash -c "cd test && make -j8 -l8 --output-sync test"
```

**Comment by jberthold** [comment]
We need to modify the required checks to merge this. Other follow-ups:
### Integration shell
- [ ] merge [PR in `K` to provide the integration shell derivation](https://github.com/runtimeverification/k/pull/3428) 
- [ ] upgrade `deps/k_release` to a version including this derivation
- [ ] _then_ modify integration tests to use the `K` version from `deps/k_release`
### Formatting PR
- [ ] activate committing the reformatted code and aborting the workflow
- [ ] PR will automatically reformat everything

**Pull request merged**